lib/srv/desktop/rdp/rdpclient/src/rdpdr/tdp.rs

```diff
@@ -192,6 +217,55 @@ pub struct SharedDirectoryCreateRequest {
     pub path: UnixPath,
 }
 
+impl SharedDirectoryCreateRequest {
+    pub fn from(req: &DeviceCreateRequest, file_type: FileType) -> SharedDirectoryCreateRequest {
+        SharedDirectoryCreateRequest {
+            completion_id: req.device_io_request.completion_id,
+            directory_id: req.device_io_request.device_id,
+            file_type,
+            path: UnixPath::from(&req.path),
+        }
+    }
+
+    /// Converts this request into a [`CGOSharedDirectoryCreateRequest`].
+    ///
+    /// Returns a tuple containing the [`CGOSharedDirectoryCreateRequest`] and a [`CString`],
+    /// which is the memory backing the [`CGOSharedDirectoryCreateRequest::path`] field.
+    /// It is the caller's responsibility to ensure that the [`CString`] lives until
+    /// the [`CGOSharedDirectoryCreateRequest::path`] is copied into Go-owned memory.
+    ///
+    /// ```
+    /// fn example() {
+    ///     let req = SharedDirectoryCreateRequest {
+    ///         completion_id: 0,
+    ///         directory_id: 0,
+    ///         file_type: FileType::File,
+    ///         path: UnixPath::from("/tmp/test.txt"),
+    ///     };
+    ///
+    ///     // using _path (as opposed to _) ensures that the CString is not dropped
+    ///     // until after the CGOSharedDirectoryCreateRequest is copied into Go-owned memory.
+    ///     let (cgo_req, _path) = req.into_cgo().unwrap();
+    ///
+    ///     copy_into_go_memory(cgo_req);
+    ///
+    ///     // _path is dropped here
+    /// }
+    /// ```
+    pub fn into_cgo(self) -> PduResult<(CGOSharedDirectoryCreateRequest, CString)> {
+        let path = self.path.to_cstring()?;
+        Ok((
```

Comment:
This is unsafe and very easy to misuse. You can wrap `CGOSharedDirectoryCreateRequest` and `CString` in struct that will only allow access through mutable ref - this will ensure that `CString` lives long enough.
For example (but of course you can create specific struct that will hide `CString` completely):
```suggestion
    pub fn into_cgo(self) -> PduResult<RefCell<(CGOSharedDirectoryCreateRequest, CString)>> {
        let path = self.path.to_cstring()?;
        Ok(RefCell::new((
```

Reply:
> This is unsafe and very easy to misuse

Valid, wdyt about this: a091305e34b45c55368792d54993f473a8185d56

Reply:
That works, thanks!